adapter/internal/ingress/client_manager.go

```diff
@@ -23,9 +23,10 @@ type connection struct {
 type ClientManager struct {
 	connector     ConnectionBuilder
 	connectionTTL time.Duration
-	connections   []unsafe.Pointer
+	connections   []*connection
```

Comment:
We are taking a global lock on all connections whenever there is a read or write for a given connection. Since all operations are on individual elements of the slice and not on the entire slice itself, this seems like a good candidate for lock striping where each index would get it's own lock.

As an example, when a thread goes to do a write for a single connection ALL readers and writers (most of which do not care about the connection being written) have to wait. Seems a bit absurd.